[ews-build.webkit.org] Cherry-pick instead of rebase PRs

JonWBedard opened the pull request:
#### 8ed8d160f642d21bf6751dbcb0b9f4e9b40543bb
<pre>
[ews-build.webkit.org] Cherry-pick instead of rebase PRs
<a href="https://bugs.webkit.org/show_bug.cgi?id=248505">https://bugs.webkit.org/show_bug.cgi?id=248505</a>
rdar://102790083

Reviewed by Ryan Haddad.

* Tools/CISupport/ews-build/steps.py:
(CheckOutPullRequest.run): Instead of checking out and rebasing the target PR, cherry-pick all commits
from the target PR onto the base commit we&apos;re testing.
(CleanGitRepo.run): Abort any in progress cherry-pick.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/257193@main">https://commits.webkit.org/257193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e160d801876a3b2752d4c39beb0cf542afaeb3f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107537 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167805 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102015 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7779 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36054 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90690 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104149 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103727 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5850 "Build was cancelled. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Pull request contains relevant changes; Skipped layout-tests; layout-tests (exception)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84674 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32987 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87719 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89430 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1265 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/97001 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1230 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22367 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6090 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2469 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2530 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41772 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -1028,10 +1028,8 @@ def run(self):
-            ['git', 'checkout', pr_branch],
-            ['git', 'config', 'merge.changelog.driver', 'perl Tools/Scripts/resolve-ChangeLogs --merge-driver -c %O %A %B'],
-            ['git', 'rebase', rebase_target_hash, pr_branch],
+            ['git', 'checkout', '-b', pr_branch],
```
Not a `git` expert, but I thought that we wanted to check out `main` here.

JonWBedard replied:
We're leaning on the existing checkout state, which was configured by `CheckOutSource` and `CheckOutSpecificRevision`. That's going to be either the base branch for builders (which will be `main` most of the time) or the `ews_revision` as passed to the tester by the builder. Somewhat confusingly, `git checkout -b <branch name>` doesn't checkout the remote version of that branch into your current local checkout, it creates a local branch named `<branch name>` from the current checkout state. We probably don't have to do this, but I wanted to keep the local checkout "using" the PR branch, for `merge-queue`, in particular, it means we don't have to change anything.

aproskuryakov replied:
Ah right, not sure how I overlooked `-b`.

Certainly makes sense locally. It's icky that a new branch is created for an existing branch, can there be conflicts in any scenarios? I don't immediately see the code that deletes the local pr_branch, and I'm not sure if pushing it still works when it's created like this. I guess I'll leave reviewing to someone who understands the overall structure of this code better.

ryanhaddad replied:
> I don't immediately see the code that deletes the local pr_branch

Deleting all local branches happens here https://github.com/WebKit/WebKit/blob/6e160d801876a3b2752d4c39beb0cf542afaeb3f/Tools/CISupport/ews-build/steps.py#L4626

ryanhaddad replied:
>I'm not sure if pushing it still works when it's created like this.

merge-queue will force push to the branch, so this should "just work"

JonWBedard replied:
>  >I'm not sure if pushing it still works when it's created like this.
> 
> 
> 
> merge-queue will force push to the branch, so this should "just work"

Right, and we have to do it this way to get a PR "merged" instead of closed, because the commit we're actually landing gets its message changed when we insert the reviewer and add the canonicalization link

JonWBedard committed:
[ews-build.webkit.org] Cherry-pick instead of rebase PRs
https://bugs.webkit.org/show_bug.cgi?id=248505
rdar://102790083

Reviewed by Ryan Haddad.

* Tools/CISupport/ews-build/steps.py:
(CheckOutPullRequest.run): Instead of checking out and rebasing the target PR, cherry-pick all commits
from the target PR onto the base commit we're testing.
(CleanGitRepo.run): Abort any in progress cherry-pick.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/257193@main
